components/patient-data/rest/src/main/java/org/phenotips/data/rest/internal/DefaultPatientsFetchResourceImpl.java

```diff
@@ -0,0 +1,222 @@
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        } catch (final QueryException ex) {
+            logger.warn("Failed to retrieve patients with external ids [{}]: {}", eids, ex.getMessage());
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        }
+    }
+
+    /**
+     * Retrieves patient entities given a list of external patient IDs.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param eids a list of external patient IDs, as strings
+     * @throws QueryException if the query fails
+     */
+    private void addEids(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final List<Object> eids) throws QueryException
+    {
+        if (CollectionUtils.isNotEmpty(eids)) {
+            collectPatientsFromEids(patientsBuilder, eids);
+        }
+    }
+
+    /**
+     * Retrieves and collects patient entities that correspond to the provided external ID.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param eids external patient IDs, as a list
+     * @throws QueryException if the query fails
+     */
+    private void collectPatientsFromEids(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final List<Object> eids) throws QueryException
+    {
+        final Query q = qm.createQuery("from doc.object(PhenoTips.PatientClass) p where p.external_id in (:eids)",
+            Query.XWQL);
+        q.bindValue("eids", eids);
+        final List<Object> patientIds = q.execute();
+        addIds(patientsBuilder, patientIds);
+    }
+
+    /**
+     * Retrieves patient entities given a list of internal patient IDs.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param ids a list of patient ids, as strings
+     */
+    private void addIds(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final List<Object> ids)
+    {
+        for (final Object id : ids) {
+            if (StringUtils.isNotBlank((String) id)) {
+                addPatientFromId(patientsBuilder, id);
+            }
+        }
+    }
+
+    /**
+     * Given the patient's internal ID, retrieves the patient entity, if it exists and if the user has view rights, and
+     * adds it to the set of patient entities.
+     *
+     * @param patientsBuilder a patient entity set builder
+     * @param id an internal patient ID
+     */
+    private void addPatientFromId(@Nonnull final ImmutableSet.Builder<PrimaryEntity> patientsBuilder,
+        @Nonnull final Object id)
+    {
+        try {
+            // Try to get the patient entity.
+            final PrimaryEntity patient = this.repository.get((String) id);
+            // If user has view rights and patient with the provided ID exists, add patient to patient set.
+            if (patient != null) {
```

Comment:
currently there is no distinction between incorrect IDs and IDs for patients which the requestor has no permissions for. This is fine, as long as that is what we want: on lines 112-118 (https://github.com/phenotips/phenotips/pull/2163/files#diff-db33e60f31a18452c5507f7605e8385dR112) the only possibl responses are OK and INTERNAL_SERVER_ERROR.

I'm not sure we want to complicate our API with more error codes, but it may be useful as well, so worth a discussion/documentation.

...in any case this is ok for version 1